code/game/objects/items/stacks/stack.dm

```diff
@@ -36,7 +36,11 @@
 /obj/item/stack/Initialize(mapload, new_amount=null , merge = TRUE)
 	. = ..()
 	if(new_amount)
-		amount = new_amount
+		if(new_amount > max_amount)
+			amount = max_amount
+			new type(loc, new_amount - max_amount, merge)
```

Comment:
I might suggest a loop here rather than recursion, though huge numbers of stacks will cause problems no matter what so it's not that important.